File: wallpaper_rotator.js
// wallpaper_rotator.js
#!/usr/bin/env node
const fs = require('fs');
const path = require('path');
const { exec, execSync } = require('child_process');
const os = require('os');

class WallpaperRotator {
    constructor() {
        this.dir = null;
        this.images = [];
        this.interval = 60;
        this.randomOrder = false;
        this.currentIndex = 0;
        this.running = false;
        this.timer = null;
        this.currentWall = null;
        this.loadState();
    }

    loadState() {
        const stateFile = path.join(os.homedir(), '.wallpaper_rotator_state.txt');
        if (fs.existsSync(stateFile)) {
            const data = fs.readFileSync(stateFile, 'utf8').split('\n');
            if (data.length >= 4) {
                this.dir = data[0] || null;
                this.interval = parseInt(data[1]) || 60;
                this.randomOrder = data[2] === 'true';
                this.currentWall = data[3] || null;
                this.scanImages();
            }
        }
    }

    saveState() {
        const stateFile = path.join(os.homedir(), '.wallpaper_rotator_state.txt');
        const content = `${this.dir || ''}\n${this.interval}\n${this.randomOrder}\n${this.currentWall || ''}`;
        fs.writeFileSync(stateFile, content);
    }

    scanImages() {
        if (!this.dir || !fs.existsSync(this.dir)) {
            this.images = [];
            return;
        }
        const exts = ['.jpg', '.jpeg', '.png', '.bmp', '.gif', '.tiff'];
        const files = fs.readdirSync(this.dir);
        this.images = files.filter(f => {
            const ext = path.extname(f).toLowerCase();
            return exts.includes(ext);
        }).map(f => path.join(this.dir, f));
    }

    setWallpaper(filePath) {
        if (!fs.existsSync(filePath)) {
            console.error(`File not found: ${filePath}`);
            return false;
        }
        const platform = os.platform();
        try {
            if (platform === 'win32') {
                execSync(`reg add "HKEY_CURRENT_USER\\Control Panel\\Desktop" /v Wallpaper /t REG_SZ /d "${filePath}" /f`);
                execSync('RUNDLL32.EXE user32.dll,UpdatePerUserSystemParameters 1 True');
            } else if (platform === 'darwin') {
                const script = `tell application "Finder" to set desktop picture to POSIX file "${filePath}"`;
                execSync(`osascript -e '${script}'`);
            } else {
                // Linux: try GNOME
                try {
                    execSync('gsettings get org.gnome.desktop.background picture-uri');
                    execSync(`gsettings set org.gnome.desktop.background picture-uri "file://${filePath}"`);
                } catch (e) {
                    execSync(`feh --bg-scale "${filePath}"`);
                }
            }
            this.currentWall = filePath;
            this.saveState();
            return true;
        } catch (err) {
            console.error(`Error setting wallpaper: ${err.message}`);
            return false;
        }
    }

    getCurrentWallpaper() {
        if (this.currentWall) return this.currentWall;
        const platform = os.platform();
        try {
            if (platform === 'darwin') {
                const out = execSync('osascript -e \'tell application "Finder" to get desktop picture\'', { encoding: 'utf8' });
                return out.trim();
            } else if (platform === 'linux') {
                const out = execSync('gsettings get org.gnome.desktop.background picture-uri', { encoding: 'utf8' });
                let uri = out.trim().replace(/^'|'$/g, '');
                if (uri.startsWith('file://')) return uri.slice(7);
            }
        } catch (e) {}
        return 'Unknown';
    }

    nextWallpaper() {
        if (this.images.length === 0) {
            console.log('No images in directory. Please add images.');
            return;
        }
        let idx;
        if (this.randomOrder) {
            idx = Math.floor(Math.random() * this.images.length);
        } else {
            this.currentIndex = (this.currentIndex + 1) % this.images.length;
            idx = this.currentIndex;
        }
        const img = this.images[idx];
        if (this.setWallpaper(img)) {
            console.log(`Wallpaper changed to: ${img}`);
        }
    }

    rotationLoop() {
        this.timer = setInterval(() => {
            this.nextWallpaper();
        }, this.interval * 1000);
    }

    start() {
        if (this.running) {
            console.log('Already running.');
            return;
        }
        if (!this.dir || !fs.existsSync(this.dir)) {
            console.log('Please set a directory with --add-dir');
            return;
        }
        this.scanImages();
        if (this.images.length === 0) {
            console.log('No images found in directory.');
            return;
        }
        this.running = true;
        this.rotationLoop();
        console.log(`Started rotation every ${this.interval} seconds.`);
        this.saveState();
    }

    stop() {
        if (this.timer) {
            clearInterval(this.timer);
            this.timer = null;
        }
        this.running = false;
        console.log('Stopped rotation.');
    }

    setDirectory(dirPath) {
        if (!fs.existsSync(dirPath) || !fs.statSync(dirPath).isDirectory()) {
            console.error(`Directory not found: ${dirPath}`);
            return;
        }
        this.dir = dirPath;
        this.scanImages();
        this.currentIndex = 0;
        this.saveState();
        console.log(`Directory set to: ${dirPath} (${this.images.length} images)`);
    }
}

const args = process.argv.slice(2);
let addDir = null, interval = null, random = false, start = false, stop = false, next = false, current = false;
for (let i = 0; i < args.length; i++) {
    switch (args[i]) {
        case '--add-dir': addDir = args[++i]; break;
        case '--interval': interval = parseInt(args[++i]); break;
        case '--random': random = true; break;
        case '--start': start = true; break;
        case '--stop': stop = true; break;
        case '--next': next = true; break;
        case '--current': current = true; break;
        case '--help':
            console.log(`Usage: node wallpaper_rotator.js --add-dir <path> [--interval <sec>] [--random] [--start|--stop|--next|--current]`);
            process.exit(0);
    }
}

const rotator = new WallpaperRotator();
if (addDir) rotator.setDirectory(addDir);
if (interval !== null) { rotator.interval = interval; rotator.saveState(); }
if (random) { rotator.randomOrder = true; rotator.saveState(); }
if (start) rotator.start();
else if (stop) rotator.stop();
else if (next) rotator.nextWallpaper();
else if (current) console.log(`Current wallpaper: ${rotator.getCurrentWallpaper()}`);

if (start) {
    // keep process alive
    process.stdin.resume();
}

File: wallpaper_rotator.go
// wallpaper_rotator.go
package main

import (
	"bufio"
	"flag"
	"fmt"
	"io/ioutil"
	"math/rand"
	"os"
	"os/exec"
	"path/filepath"
	"runtime"
	"strconv"
	"strings"
	"time"
)

type WallpaperRotator struct {
	dir          string
	images       []string
	interval     int
	randomOrder  bool
	currentIndex int
	running      bool
	currentWall  string
	stopChan     chan bool
}

func NewWallpaperRotator() *WallpaperRotator {
	w := &WallpaperRotator{
		interval:     60,
		randomOrder:  false,
		currentIndex: 0,
		running:      false,
		stopChan:     make(chan bool),
	}
	w.loadState()
	return w
}

func (w *WallpaperRotator) loadState() {
	stateFile := filepath.Join(os.Getenv("HOME"), ".wallpaper_rotator_state.txt")
	data, err := ioutil.ReadFile(stateFile)
	if err != nil {
		return
	}
	lines := strings.Split(string(data), "\n")
	if len(lines) >= 4 {
		w.dir = lines[0]
		interval, _ := strconv.Atoi(lines[1])
		w.interval = interval
		w.randomOrder = lines[2] == "true"
		w.currentWall = lines[3]
		w.scanImages()
	}
}

func (w *WallpaperRotator) saveState() {
	stateFile := filepath.Join(os.Getenv("HOME"), ".wallpaper_rotator_state.txt")
	content := fmt.Sprintf("%s\n%d\n%t\n%s\n", w.dir, w.interval, w.randomOrder, w.currentWall)
	ioutil.WriteFile(stateFile, []byte(content), 0644)
}

func (w *WallpaperRotator) scanImages() {
	if w.dir == "" {
		return
	}
	exts := []string{".jpg", ".jpeg", ".png", ".bmp", ".gif", ".tiff"}
	files, _ := ioutil.ReadDir(w.dir)
	w.images = []string{}
	for _, f := range files {
		if !f.IsDir() {
			ext := strings.ToLower(filepath.Ext(f.Name()))
			for _, e := range exts {
				if ext == e {
					w.images = append(w.images, filepath.Join(w.dir, f.Name()))
					break
				}
			}
		}
	}
}

func (w *WallpaperRotator) setWallpaper(path string) bool {
	if _, err := os.Stat(path); os.IsNotExist(err) {
		fmt.Printf("File not found: %s\n", path)
		return false
	}
	var cmd *exec.Cmd
	switch runtime.GOOS {
	case "windows":
		cmd = exec.Command("cmd", "/c", "reg", "add", "HKEY_CURRENT_USER\\Control Panel\\Desktop", "/v", "Wallpaper", "/t", "REG_SZ", "/d", path, "/f")
		cmd.Run()
		cmd = exec.Command("cmd", "/c", "RUNDLL32.EXE", "user32.dll,UpdatePerUserSystemParameters", "1", "True")
	case "darwin":
		script := fmt.Sprintf(`tell application "Finder" to set desktop picture to POSIX file "%s"`, path)
		cmd = exec.Command("osascript", "-e", script)
	default: // Linux
		// Try GNOME
		if _, err := exec.Command("gsettings", "get", "org.gnome.desktop.background", "picture-uri").Output(); err == nil {
			cmd = exec.Command("gsettings", "set", "org.gnome.desktop.background", "picture-uri", "file://"+path)
		} else {
			cmd = exec.Command("feh", "--bg-scale", path)
		}
	}
	err := cmd.Run()
	if err != nil {
		fmt.Printf("Error setting wallpaper: %v\n", err)
		return false
	}
	w.currentWall = path
	w.saveState()
	return true
}

func (w *WallpaperRotator) getCurrentWallpaper() string {
	if w.currentWall != "" {
		return w.currentWall
	}
	// Try to detect
	switch runtime.GOOS {
	case "darwin":
		out, _ := exec.Command("osascript", "-e", `tell application "Finder" to get desktop picture`).Output()
		return strings.TrimSpace(string(out))
	case "linux":
		out, _ := exec.Command("gsettings", "get", "org.gnome.desktop.background", "picture-uri").Output()
		uri := strings.Trim(string(out), "'\n")
		if strings.HasPrefix(uri, "file://") {
			return uri[7:]
		}
	}
	return "Unknown"
}

func (w *WallpaperRotator) nextWallpaper() {
	if len(w.images) == 0 {
		fmt.Println("No images in directory. Please add images.")
		return
	}
	var idx int
	if w.randomOrder {
		idx = rand.Intn(len(w.images))
	} else {
		w.currentIndex = (w.currentIndex + 1) % len(w.images)
		idx = w.currentIndex
	}
	path := w.images[idx]
	if w.setWallpaper(path) {
		fmt.Printf("Wallpaper changed to: %s\n", path)
	}
}

func (w *WallpaperRotator) rotationLoop() {
	ticker := time.NewTicker(time.Duration(w.interval) * time.Second)
	defer ticker.Stop()
	for {
		select {
		case <-ticker.C:
			w.nextWallpaper()
		case <-w.stopChan:
			return
		}
	}
}

func (w *WallpaperRotator) start() {
	if w.running {
		fmt.Println("Already running.")
		return
	}
	if w.dir == "" {
		fmt.Println("Please set a directory with --add-dir")
		return
	}
	w.scanImages()
	if len(w.images) == 0 {
		fmt.Println("No images found in directory.")
		return
	}
	w.running = true
	go w.rotationLoop()
	fmt.Printf("Started rotation every %d seconds.\n", w.interval)
	w.saveState()
}

func (w *WallpaperRotator) stop() {
	if !w.running {
		fmt.Println("Not running.")
		return
	}
	w.running = false
	w.stopChan <- true
	fmt.Println("Stopped rotation.")
}

func (w *WallpaperRotator) setDirectory(path string) {
	if _, err := os.Stat(path); os.IsNotExist(err) {
		fmt.Printf("Directory not found: %s\n", path)
		return
	}
	w.dir = path
	w.scanImages()
	w.currentIndex = 0
	w.saveState()
	fmt.Printf("Directory set to: %s (%d images)\n", path, len(w.images))
}

func main() {
	var (
		addDir   = flag.String("add-dir", "", "Directory with wallpapers")
		interval = flag.Int("interval", 0, "Rotation interval in seconds")
		random   = flag.Bool("random", false, "Random order")
		start    = flag.Bool("start", false, "Start rotation")
		stop     = flag.Bool("stop", false, "Stop rotation")
		next     = flag.Bool("next", false, "Force next wallpaper")
		current  = flag.Bool("current", false, "Show current wallpaper")
	)
	flag.Parse()

	w := NewWallpaperRotator()

	if *addDir != "" {
		w.setDirectory(*addDir)
	}
	if *interval > 0 {
		w.interval = *interval
		w.saveState()
	}
	if *random {
		w.randomOrder = true
		w.saveState()
	}
	if *start {
		w.start()
	} else if *stop {
		w.stop()
	} else if *next {
		w.nextWallpaper()
	} else if *current {
		cur := w.getCurrentWallpaper()
		fmt.Printf("Current wallpaper: %s\n", cur)
	} else {
		flag.Usage()
	}

	if *start {
		// keep alive
		select {}
	}
}
